src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/LspFileBasedProgramWorkspace.cs

```diff
@@ -0,0 +1,323 @@
+            var document1 = this.CurrentSolution.GetDocument(documentId1);
+            return document1;
+        }
+
+        logger?.LogDebug("Projects count: {0}", loadedProjectInfos.Length);
+
+        var projectSystemName = $"{documentFilePath} (${loadedProjectInfo.TargetFramework})";
+        var projectCreationInfo = new ProjectSystemProjectCreationInfo
+        {
+            AssemblyName = projectSystemName,
+            FilePath = documentFilePath,
+            CompilationOutputAssemblyFilePath = loadedProjectInfo.IntermediateOutputFilePath
+        };
+
+        var ProjectSystemProjectFactory = new ProjectSystemProjectFactory(
+            this, _fileChangeWatcher, static (_, _) => Task.CompletedTask, _ => { },
+            CancellationToken.None); // TODO: do we need to introduce a shutdown cancellation token for this?
+        // workspace.ProjectSystemProjectFactory = ProjectSystemProjectFactory;
+
+        var ProjectSystemHostInfo = new ProjectSystemHostInfo(
+            DynamicFileInfoProviders: [.. _dynamicFileInfoProviders],
+            new HostDiagnosticAnalyzerProvider(razorSourceGenerator: null),
+            AnalyzerAssemblyRedirectors: []);
+
+        var projectSystemProject = await ProjectSystemProjectFactory.CreateAndAddToWorkspaceAsync(
+            projectSystemName,
+            loadedProjectInfo.Language,
+            projectCreationInfo,
+            ProjectSystemHostInfo);
+
+        // TODO: restore. see LanguageServerProjectSystem.cs
+        // TODO: not retaining the project, e.g. returning same cached project/document on repeated calls, is likely messing things up
+
+        var loadedProject = new LoadedProject(projectSystemProject, Services.SolutionServices, _fileChangeWatcher, _projectTargetFrameworkManager);
+        loadedProject.NeedsReload += (_, _) => _projectsToLoadAndReload.AddWork(pair);
+        existingProjects.Add(loadedProject);
+
+        var (targetTelemetryInfo, targetNeedsRestore) = await loadedProject.UpdateWithNewProjectInfoAsync(loadedProjectInfo, iLogger);
+
+        var documentId = this.CurrentSolution.GetDocumentIdsWithFilePath(documentFilePath).First();
+        var document = this.CurrentSolution.GetDocument(documentId);
+        return document;
+    }
+
+    private string MakeVirtualProjectContent(string documentFilePath, SourceText documentText)
+    {
+        return $"""
+            <Project Sdk="Microsoft.NET.Sdk">
+
+                <PropertyGroup>
+                    <OutputType>Exe</OutputType>
+                    <TargetFramework>net9.0</TargetFramework>
+                    <ImplicitUsings>enable</ImplicitUsings>
+                    <Nullable>enable</Nullable>
+                    <Features>$(Features);FileBasedProgram</Features>
+                    <EnableDefaultCompileItems>false</EnableDefaultCompileItems>
+                </PropertyGroup>
+
+                <ItemGroup>
+                    <Compile Include="{documentFilePath}" />
```

Comment:
this is not at all safe.  (effectively, this is like sql injection).  We should not be manually making xml.  We should use an xml creation lib to get to this to happen.

Reply:
This is going to be deleted and replaced with usage of API described in #78159.